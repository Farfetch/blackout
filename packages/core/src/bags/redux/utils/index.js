/**
 * Bags utils.
 *
 * @module bags/utils
 * @category Bags
 * @subcategory Utils
 */
export { default as createBagItemHash } from './createBagItemHash';
export { default as buildBagItem } from './buildBagItem';
export { default as areBagItemsIdentical } from './areBagItemsIdentical';
export { default as bagOperationChangeTypes } from './bagOperationChangeTypes';
