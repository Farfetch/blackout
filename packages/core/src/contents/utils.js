/**
 * Contents utils.
 *
 * @module contents/utils
 * @category Contents
 * @subcategory Utils
 */

import get from 'lodash/get';
import isEmpty from 'lodash/isEmpty';

/**
 * Constant that represent all possible static values to apply to an environment code variable.
 *
 * @type {object}
 */
export const ENVIRONMENT_CODES = {
  LIVE: 'live',
  PREVIEW: 'preview',
};

/**
 * Constant that represent all possible genders.
 *
 * @type {object}
 */
export const GENDER = {
  0: 'Woman',
  1: 'Man',
  2: 'Unisex',
  3: 'Kids',
};

/**
 * Constant that represent the price types available.
 *
 * @type {object}
 */
export const PRICETYPE = {
  0: 'Full Price',
  1: 'Sale',
  2: 'Private Sale',
};

/**
 * Generate a ranking number for each commerce page.
 *
 * @function
 *
 * @param {object} metadata - Metadata object with custom attributes.
 * @returns {number} Ranking number for a specific commerce page.
 *
 * @example
 * const pageRanking = getPageRanking({ custom: { type: "listing" } });
 * Result of pageRanking === 0;
 */
export const getPageRanking = metadata => {
  let ranking = 0;
  const metadataObj =
    typeof metadata.custom === 'string'
      ? JSON.parse(metadata.custom)
      : metadata.custom;
  const { gender, brand, category, priceType, id } = metadataObj;

  // Return default ranking number when no data was returned from each commerce property.
  // We can't only check if it's empty from 'metadata.custom' because it could have other properties inside.
  if (!gender && !brand && !category && !priceType && !id) {
    return ranking;
  }

  const rankingGender = 10;
  const rankingBrand = 100;
  const rankingCategory = 10000;
  const rankingPriceType = 1000000000;
  const rankingId = 10000000000;

  Object.entries(metadataObj).forEach(entry => {
    const [key, value] = entry;

    switch (key) {
      case 'gender':
        ranking += !isEmpty(value) && rankingGender;
        break;
      case 'brand':
        ranking += !isEmpty(value) && rankingBrand;
        break;
      case 'category':
        ranking += !isEmpty(value) && rankingCategory;
        break;
      case 'priceType':
        ranking += !isEmpty(value) && rankingPriceType;
        break;
      case 'id':
        ranking += !isEmpty(value) && rankingId;
        break;

      default:
        return ranking;
    }
  });

  return ranking;
};

/**
 * Method to check each page ranking and return a list of ranked commerce pages ordered by the better one.
 *
 * @function
 *
 * @param {object} result - Commerce page payload result.
 * @returns {object[]} - List of commerce pages ranked.
 *
 * @example
 * const getCommercePagesRanked = getCommercePagesRanked({ entries: [...pages] });
 * Result of getCommercePagesRanked === { entries: [commercePagesWithRankProperty] };
 */
const getCommercePagesRanked = result => {
  const commercePagesWithRank = result.entries.map(page => ({
    ...page,
    ranking: getPageRanking(page.metadata),
  }));

  return commercePagesWithRank.sort((p, c) => c.ranking - p.ranking);
};

/**
 * Method to check each page ranking and return the better one.
 *
 * @function
 *
 * @param {object} result - Commerce page payload result.
 * @returns {object} Selected page with better ranking.
 *
 * @example
 * const defaultStrategy = getDefaultStrategy({ entries: [...pages] });
 * Result of defaultStrategy === { entries: [...bestRankedPage] };
 */
export const getDefaultStrategy = result => {
  const bestRankedPage = getCommercePagesRanked(result)[0];

  return {
    number: 1,
    totalPages: 1,
    totalItems: 1,
    entries: [bestRankedPage],
  };
};

/**
 * Method to return the main ranked commerce page with all components from other pages without duplicates.
 *
 * @function
 *
 * @param {object} result - Commerce page payload result.
 * @returns {object} Selected page with merged components.
 *
 * @example
 * const mergeStrategy = getMergeStrategy({ entries: [...pages] });
 * Result of mergeStrategy === { entries: [...mergedPages] };
 */
export const getMergeStrategy = result => {
  const rankedCommercePages = getCommercePagesRanked(result);
  // Merge components inside the selected commerce page
  const mergedComponents = [];

  rankedCommercePages.forEach(content =>
    content.components.forEach(component => {
      if (
        !mergedComponents.some(
          mergedComponent =>
            mergedComponent.customType === component.customType,
        )
      )
        mergedComponents.push(component);
    }),
  );

  return {
    number: 1,
    totalItems: 1,
    totalPages: 1,
    entries: [{ ...rankedCommercePages[0], components: mergedComponents }],
  };
};

/**
 * Method to calculate the ranking for commerce pages, according to the selected strategy
 * and returning only the selected one.
 *
 * @function
 *
 * @param {object} result - Commerce page payload result.
 * @param {string} [strategy] - The selected ranking strategy (E.g. 'default | merge').
 * @returns {object} Selected page with best ranking.
 *
 * @example
 * const rankedPage = getRankedCommercePage({ entries: [...pages] }, 'merge');
 * Result of rankedPage === { entries: [...mergedPages] };
 */
export const getRankedCommercePage = (result, strategy) => {
  result.entries.forEach(commerce => {
    if (typeof commerce?.metadata?.custom === 'string') {
      console.warn(`[Commerce Pages]: Seems you are trying to fetch legacy commerce page.
      Try to republish the commerce page to update data.`);
    }
  });

  switch (strategy) {
    case 'merge':
      return getMergeStrategy(result);
    default:
      return getDefaultStrategy(result);
  }
};

/**
 * Build a hash with query object received to identify contentGroups when searching for contents.
 *
 * @function
 *
 * @param {object} query - Object with query parameters applied to search contents.
 * @param {string | string[]} [query.codes] - List of codes that representing the content code (about-us|today-news|header|productId...).
 * @param {object} [query.contentTypeCode] - The content type unique code (page|post|menu|pages|posts|widgets|waterproof...).
 * @param {number} [query.page] - Number of the page to get, starting at 1. The default is 1.
 * @param {number} [query.pageSize] - Size of each page, as a number between 1 and 180. The default is 60.
 *
 * @returns {string} - Hash built to identify a content group.
 *
 * @example
 * const contentHash = buildContentGroupHash({ codes: 'about', contentTypeCode: 'pages'});
 * Result of contentGroupHash === 'pages!about';
 */
export const buildContentGroupHash = query => {
  if (
    isEmpty(query) ||
    (!get(query, 'contentTypeCode') && !get(query, 'codes'))
  ) {
    return '';
  }

  const contentType = get(query, 'contentTypeCode', 'all');
  const codes = get(query, 'codes', 'all');
  const page = get(query, 'page', 1);
  const pageSize = get(query, 'pageSize', '');
  const pagesQuery = pageSize && `,${pageSize}`;

  return `${contentType}!${codes}!${page}${pagesQuery}`;
};

/**
 * Build a hash with query object received to identify the type and page of the metadata selected.
 *
 * @function
 *
 * @param {object} query - Object with query parameters applied to search for metadata.
 * @param {object} query.path - The pathname of the location.
 * @param {object} [query.pageType] - The type of the page we are searching (pages|stories...).
 *
 * @returns {string} - Hash built to identify the metadata for a specific page and pageType.
 *
 * @example
 * const pathname = buildSEOPathname({ path: '/about', pageType: 'pages'});
 * Result of pathname === 'pages!/about';
 */
export const buildSEOPathname = query => {
  if (isEmpty(query) || !get(query, 'path')) {
    return '';
  }

  const pageType = get(query, 'pageType', 'all');

  return `${pageType}!${query.path}`;
};

/**
 * Strip the slug to remove the subfolder and json=true.
 *
 * @param {string} slug - The slug of url.
 * @param {string} subfolder - The subforlder of url.
 *
 * @returns {string} - The slug without subfolder / json=true.
 */
export const stripSlugSubfolderJsonTrue = (slug, subfolder) => {
  let slugMetada = slug;

  if (subfolder !== '/') {
    slugMetada = slug.replace(subfolder, '');
  }

  // Remove json=true from slug
  return slugMetada.replace('?json=true', '').replace('&json=true', '');
};
