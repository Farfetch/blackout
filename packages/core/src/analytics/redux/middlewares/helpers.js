/**
 * @module
 * @private
 */

import {
  getBrand as getBrandSelector,
  getCategory as getCategorySelector,
} from '../../../entities/redux/selectors';
import { logger } from '../../utils';
import get from 'lodash/get';

/**
 * Returns the different product categories separated with a `/`.
 *
 * @param {object} state - The current application state.
 * @param {object} product - The product object.
 *
 * @returns {string} String with all category names joined with a '/' character.
 */
export const getCategory = (state, product) =>
  get(product, 'categories', [])
    .map(id => get(getCategorySelector(state, id), 'name'))
    .join('/');

/**
 * Returns the product brand name.
 *
 * @param {object} state - The current application state.
 * @param {object} product - The product object.
 *
 * @returns {(string|undefined)} The brand name for the product.
 */
export const getBrand = (state, product) =>
  get(getBrandSelector(state, get(product, 'brand')), 'name');

/**
 * Returns the product variant (color).
 *
 * @param {object} product - The product object.
 *
 * @returns {(string|undefined)} The product variant.
 */
export const getVariant = product => {
  const mainVariant = get(product, 'colors', []).find(color =>
    get(color, 'tags', []).some(tag => tag === 'DesignerColor'),
  );

  return get(mainVariant, 'color.name');
};

/**
 * Returns the product size by the selected size id.
 *
 * @param {object} product - The product object.
 * @param {number} sizeId - The selected size ID.
 *
 * @returns {(string|undefined)} The size name.
 */
export const getSize = (product, sizeId) =>
  get(
    get(product, 'sizes', []).find(size => get(size, 'id') === sizeId),
    'name',
  );

/**
 * Returns the currency code passed to analytics via `useContext()`.
 *
 * @param {Analytics} analyticsInstance - Analytics instance.
 *
 * @returns {(string|undefined)} The currency from analytics's context.
 */
export const getCurrency = async analyticsInstance => {
  const currency = await analyticsInstance.context('currencyCode');

  if (!currency) {
    logger.error(
      'Track event failed. Make sure to set `currencyCode` via `analytics.useContext()`.',
    );
  }

  return currency;
};

/**
 * Calculates the price discount absolute value by subtracting the price
 * with discount value from the price without discount value.
 * Using the price.discount.rate or even the price.discount.includingTaxes
 * was considered but they introduce additional complexity as they might
 * not exist in the store or cause rounding errors in the case of the price.discount.rate
 * as it is a percentage.
 *
 * @param {number} priceWithDiscount - The price value with discount.
 * @param {number} priceWithoutDiscount - The price value without discount.
 *
 * @returns {number} The absolute discount value or 0 if either priceWithDiscount or priceWithoutDiscount are not numbers.
 */
export const calculatePriceDiscount = (
  priceWithDiscount,
  priceWithoutDiscount,
) => {
  return typeof priceWithDiscount === 'number' &&
    typeof priceWithoutDiscount === 'number'
    ? priceWithoutDiscount - priceWithDiscount
    : 0;
};
