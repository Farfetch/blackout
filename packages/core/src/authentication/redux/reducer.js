/**
 * @module authentication/reducer
 * @category Authentication
 * @subcategory Reducer
 */

import * as actionTypes from './actionTypes';
import { combineReducers } from 'redux';
import { reducerFactory } from '../../helpers/redux';

const INITIAL_STATE = {
  error: null,
  id: null,
  isLoading: false,
  login: {
    error: null,
    isLoading: false,
  },
  logout: {
    error: null,
    isLoading: false,
  },
  register: {
    error: null,
    isLoading: false,
  },
  changePassword: {
    error: null,
    isLoading: false,
  },
  recoverPassword: {
    error: null,
    isLoading: false,
  },
  resetPassword: {
    error: null,
    isLoading: false,
  },
  validateEmail: {
    error: null,
    isLoading: false,
  },
  // TODO: Remove in version 2.0
  refreshToken: {
    error: null,
    isLoading: false,
  },
  refreshEmailToken: {
    error: null,
    isLoading: false,
  },
};

const error = (state = INITIAL_STATE.error, action = {}) => {
  switch (action.type) {
    case actionTypes.LOGIN_FAILURE:
    case actionTypes.LOGOUT_FAILURE:
    case actionTypes.REGISTER_FAILURE:
    case actionTypes.PASSWORD_CHANGE_FAILURE:
    case actionTypes.PASSWORD_RECOVER_FAILURE:
    case actionTypes.PASSWORD_RESET_FAILURE:
      return action.payload.error;
    case actionTypes.AUTHENTICATION_RESET:
    case actionTypes.LOGIN_REQUEST:
    case actionTypes.LOGOUT_REQUEST:
    case actionTypes.PASSWORD_CHANGE_REQUEST:
    case actionTypes.PASSWORD_RECOVER_REQUEST:
    case actionTypes.PASSWORD_RESET_REQUEST:
    case actionTypes.REGISTER_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const id = (state = INITIAL_STATE.id, action = {}) => {
  switch (action.type) {
    case actionTypes.LOGIN_SUCCESS:
    case actionTypes.REGISTER_SUCCESS:
      return action.payload.result;
    case actionTypes.LOGOUT_SUCCESS:
    case actionTypes.AUTHENTICATION_RESET:
      return INITIAL_STATE.id;
    default:
      return state;
  }
};

const isLoading = (state = INITIAL_STATE.isLoading, action = {}) => {
  switch (action.type) {
    case actionTypes.LOGIN_REQUEST:
    case actionTypes.LOGOUT_REQUEST:
    case actionTypes.REGISTER_REQUEST:
    case actionTypes.PASSWORD_CHANGE_REQUEST:
    case actionTypes.PASSWORD_RECOVER_REQUEST:
    case actionTypes.PASSWORD_RESET_REQUEST:
      return true;
    case actionTypes.LOGIN_SUCCESS:
    case actionTypes.LOGOUT_SUCCESS:
    case actionTypes.REGISTER_SUCCESS:
    case actionTypes.PASSWORD_CHANGE_SUCCESS:
    case actionTypes.PASSWORD_RECOVER_SUCCESS:
    case actionTypes.PASSWORD_RESET_SUCCESS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

export const login = reducerFactory('LOGIN', INITIAL_STATE.login, actionTypes);
export const logout = reducerFactory(
  'LOGOUT',
  INITIAL_STATE.logout,
  actionTypes,
);
export const register = reducerFactory(
  'REGISTER',
  INITIAL_STATE.register,
  actionTypes,
);
export const changePassword = reducerFactory(
  'PASSWORD_CHANGE',
  INITIAL_STATE.changePassword,
  actionTypes,
);
export const resetPassword = reducerFactory(
  'PASSWORD_RESET',
  INITIAL_STATE.resetPassword,
  actionTypes,
);
export const recoverPassword = reducerFactory(
  'PASSWORD_RECOVER',
  INITIAL_STATE.recoverPassword,
  actionTypes,
);
export const validateEmail = reducerFactory(
  'VALIDATE_EMAIL',
  INITIAL_STATE.validateEmail,
  actionTypes,
);
export const refreshToken = reducerFactory(
  'REFRESH_TOKEN',
  INITIAL_STATE.refreshToken,
  actionTypes,
);
export const refreshEmailToken = reducerFactory(
  'REFRESH_EMAIL_TOKEN',
  INITIAL_STATE.refreshEmailToken,
  actionTypes,
);

export const entitiesMapper = {
  [actionTypes.LOGOUT_SUCCESS]: state => {
    return {
      ...state,
      user: {},
    };
  },
};

export const getError = state => state.error;
export const getIsLoading = state => state.isLoading;
export const getId = state => state.id;
export const getLogin = state => state.login;
export const getLogout = state => state.logout;
export const getRegister = state => state.register;
export const getChangePassword = state => state.changePassword;
export const getResetPassword = state => state.resetPassword;
export const getRecoverPassword = state => state.recoverPassword;
export const getValidateEmail = state => state.validateEmail;
export const getRefreshEmailToken = state => state.refreshEmailToken;
export const getRefreshToken = state => state.refreshToken; // @TODO: Remove in version 2.0

/**
 * Reducer for authentication state.
 *
 * @function authenticationReducer
 * @static
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
export default combineReducers({
  error,
  id,
  isLoading,
  login,
  logout,
  register,
  changePassword,
  resetPassword,
  recoverPassword,
  validateEmail,
  refreshToken, // @TODO: To remove in the next version 2.0
  refreshEmailToken,
});
