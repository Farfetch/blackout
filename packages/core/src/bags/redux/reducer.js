/**
 * @module bags/reducer
 * @category Bags
 * @subcategory Reducer
 */
import * as actionTypes from './actionTypes';
import { combineReducers } from 'redux';
import { createMergedObject } from '../../helpers/redux';

const INITIAL_STATE = {
  error: null,
  id: null,
  isLoading: false,
  bagItems: {
    error: {},
    isLoading: {},
  },
  bagOperations: {
    error: {},
    isLoading: {},
  },
};

const error = (
  state = INITIAL_STATE.error,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.ADD_ITEM_TO_BAG_FAILURE:
    case actionTypes.GET_BAG_FAILURE:
      return action.payload.error;
    case actionTypes.ADD_ITEM_TO_BAG_REQUEST:
    case actionTypes.DELETE_BAG_ITEM_REQUEST:
    case actionTypes.GET_BAG_REQUEST:
    case actionTypes.UPDATE_BAG_ITEM_REQUEST:
    case actionTypes.GET_BAG_OPERATION_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const id = (
  state = INITIAL_STATE.id,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_BAG_SUCCESS:
      return action.payload.result;
    default:
      return state;
  }
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.ADD_ITEM_TO_BAG_REQUEST:
    case actionTypes.GET_BAG_REQUEST:
      return true;
    case actionTypes.ADD_ITEM_TO_BAG_FAILURE:
    case actionTypes.ADD_ITEM_TO_BAG_SUCCESS:
    case actionTypes.GET_BAG_FAILURE:
    case actionTypes.GET_BAG_SUCCESS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

const bagItems = (
  state = INITIAL_STATE.bagItems,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.DELETE_BAG_ITEM_REQUEST:
    case actionTypes.UPDATE_BAG_ITEM_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.bagItemId]: true,
        },
        error: {
          ...state.error,
          [action.payload.bagItemId]: null,
        },
      };
    case actionTypes.DELETE_BAG_ITEM_SUCCESS:
    case actionTypes.UPDATE_BAG_ITEM_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.bagItemId]: false,
        },
      };
    case actionTypes.DELETE_BAG_ITEM_FAILURE:
    case actionTypes.UPDATE_BAG_ITEM_FAILURE:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.bagItemId]: false,
        },
        error: {
          ...state.error,
          [action.payload.bagItemId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

const bagOperations = (
  state = INITIAL_STATE.bagOperations,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_BAG_OPERATION_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.bagOperationId]: true,
        },
        error: {
          ...state.error,
          [action.payload.bagOperationId]: null,
        },
      };
    case actionTypes.GET_BAG_OPERATION_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.bagOperationId]: false,
        },
      };
    case actionTypes.GET_BAG_OPERATION_FAILURE:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.bagOperationId]: false,
        },
        error: {
          ...state.error,
          [action.payload.bagOperationId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

export const entitiesMapper = {
  [actionTypes.DELETE_BAG_ITEM_SUCCESS]: (
    state,
    /* istanbul ignore next */ action = {},
  ) => {
    const id = action.payload.result;
    const bag = action.payload.entities.bag[id];
    const newState = createMergedObject(state, action.payload.entities);

    newState.bag[id] = bag;

    return newState;
  },
  [actionTypes.RESET_BAG_ENTITIES]: state => {
    const { bag, bagItems, bagOperations, ...rest } = state;

    return rest;
  },
  [actionTypes.RESET_BAG_OPERATIONS]: state => {
    const { bagOperations, ...rest } = state;

    return rest;
  },
};

export const getError = state => state.error;
export const getId = state => state.id;
export const getIsLoading = state => state.isLoading;
export const getIsBagItemLoading = state => state.bagItems.isLoading;
export const getItemError = state => state.bagItems.error;
export const getIsBagOperationLoading = state => state.bagOperations.isLoading;
export const getBagOperationError = state => state.bagOperations.error;

const reducer = combineReducers({
  error,
  id,
  isLoading,
  bagItems,
  bagOperations,
});

/**
 * Reducer for bag state.
 *
 * @function bagReducer
 * @static
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
export default (state, action = {}) => {
  if (action.type === actionTypes.RESET_BAG_STATE) {
    const fieldsToReset = action.payload.fieldsToReset;

    if (!fieldsToReset) {
      return INITIAL_STATE;
    } else {
      const reducerFn = (acc, field) => {
        if (state.bagItems[field]) {
          return {
            ...acc,
            [field]: INITIAL_STATE[field],
            bagItems: {
              ...acc.bagItems,
              [field]: INITIAL_STATE.bagItems[field],
            },
          };
        }

        return { ...acc, [field]: INITIAL_STATE[field] };
      };

      return fieldsToReset.reduce(reducerFn, state);
    }
  }

  return reducer(state, action);
};
