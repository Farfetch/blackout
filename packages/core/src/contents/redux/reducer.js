/**
 * @module contents/reducer
 * @category Contents
 * @subcategory Reducer
 */

import * as actionTypes from './actionTypes';
import { combineReducers } from 'redux';

export const INITIAL_STATE = {
  error: {},
  isLoading: {},
  contentTypes: {
    error: {},
    isLoading: {},
    result: null,
  },
  metadata: {
    error: {},
    isLoading: {},
    result: null,
  },
};

const error = (state = INITIAL_STATE.error, action) => {
  switch (action.type) {
    case actionTypes.GET_CONTENT_REQUEST:
    case actionTypes.GET_COMMERCE_PAGES_REQUEST:
    case actionTypes.GET_CONTENT_PAGES_REQUEST:
      return {
        ...state,
        [action.payload.hash]: null,
      };
    case actionTypes.GET_CONTENT_FAILURE:
    case actionTypes.GET_COMMERCE_PAGES_FAILURE:
    case actionTypes.GET_CONTENT_PAGES_FAILURE:
      return {
        ...state,
        [action.payload.hash]: action.payload.error,
      };
    default:
      return state;
  }
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_CONTENT_REQUEST:
    case actionTypes.GET_COMMERCE_PAGES_REQUEST:
    case actionTypes.GET_CONTENT_PAGES_REQUEST:
      return {
        ...state,
        [action.payload.hash]: true,
      };
    case actionTypes.GET_CONTENT_SUCCESS:
    case actionTypes.GET_COMMERCE_PAGES_SUCCESS:
    case actionTypes.GET_CONTENT_PAGES_SUCCESS:
      return {
        ...state,
        [action.payload.hash]: false,
      };
    case actionTypes.GET_CONTENT_FAILURE:
    case actionTypes.GET_COMMERCE_PAGES_FAILURE:
    case actionTypes.GET_CONTENT_PAGES_FAILURE:
      return {
        ...state,
        [action.payload.hash]: undefined,
      };
    default:
      return state;
  }
};

const contentTypes = (
  state = INITIAL_STATE.contentTypes,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_CONTENT_TYPES_REQUEST:
      return {
        isLoading: true,
        error: INITIAL_STATE.contentTypes.error,
      };
    case actionTypes.GET_CONTENT_TYPES_SUCCESS:
      return {
        result: action.payload,
        isLoading: false,
      };
    case actionTypes.GET_CONTENT_TYPES_FAILURE:
      return {
        isLoading: false,
        error: action.payload.error,
      };
    default:
      return state;
  }
};

const metadata = (
  state = INITIAL_STATE.metadata,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_SEO_REQUEST:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.pathname]: true,
        },
        error: {
          ...state.error,
          [action.payload.pathname]: null,
        },
      };
    case actionTypes.GET_SEO_SUCCESS:
      return {
        ...state,
        result: {
          ...state.result,
          [action.payload.pathname]: action.payload.result,
        },
        isLoading: {
          ...state.isLoading,
          [action.payload.pathname]: false,
        },
      };
    case actionTypes.GET_SEO_FAILURE:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.pathname]: undefined,
        },
        error: {
          ...state.error,
          [action.payload.pathname]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

export const getError = state => state.error;
export const getIsLoading = state => state.isLoading;
export const getContentType = state => state.contentTypes.result;
export const getSEOmetadata = state => state.metadata;

const reducers = combineReducers({
  error,
  isLoading,
  contentTypes,
  metadata,
});

/**
 * Reducer for contents state.
 *
 * @function contentsReducer
 * @static
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
export default (state, action = {}) => {
  if (action.type === actionTypes.RESET_CONTENTS) {
    return reducers(INITIAL_STATE, action);
  }

  return reducers(state, action);
};
