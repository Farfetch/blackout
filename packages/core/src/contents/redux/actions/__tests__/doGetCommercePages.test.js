import {
  commercePagesQuery,
  expectedCommercePagesNormalizedPayload,
  mockCommercePages,
  mockCommercePagesWithPages,
} from 'tests/__fixtures__/contents';
import { doGetCommercePages } from '../';
import { getRankedCommercePage } from '../../../utils';
import { mockStore } from '../../../../../tests';
import { warnDeprecatedMethod } from '../../../../helpers';
import find from 'lodash/find';
import reducer, { actionTypes } from '../../';

jest.mock('../../../utils', () => ({
  buildContentGroupHash: () => 'commerce_pages!woman',
  buildSEOPathname: jest.fn(),
  getRankedCommercePage: jest.fn(),
}));

jest.mock('../../../../helpers', () => ({
  ...jest.requireActual('../../../../helpers'),
  warnDeprecatedMethod: jest.fn(),
}));

const commercePagesMockStore = (state = {}) =>
  mockStore({ contents: reducer() }, state);

const normalizeSpy = jest.spyOn(require('normalizr'), 'normalize');
const expectedConfig = undefined;
let store;

describe('doGetCommercePages() action creator', () => {
  const getCommercePages = jest.fn();
  const action = doGetCommercePages(getCommercePages, 'woman');

  beforeEach(() => {
    jest.clearAllMocks();
    store = commercePagesMockStore();
  });

  it('should create the correct actions for when the get commerce pages procedure fails', async () => {
    const expectedError = new Error('Get commerce pages error');

    getCommercePages.mockRejectedValueOnce(expectedError);

    expect.assertions(4);

    try {
      await store.dispatch(action(commercePagesQuery));
    } catch (error) {
      expect(error).toBe(expectedError);
      expect(getCommercePages).toHaveBeenCalledTimes(1);
      expect(getCommercePages).toHaveBeenCalledWith(
        commercePagesQuery,
        expectedConfig,
      );
      expect(store.getActions()).toEqual(
        expect.arrayContaining([
          expect.objectContaining(
            {
              type: actionTypes.GET_COMMERCE_PAGES_FAILURE,
            },
            {
              payload: { error: expectedError },
              type: actionTypes.GET_COMMERCE_PAGES_FAILURE,
            },
          ),
        ]),
      );
    }
  });

  it('should create the correct actions for when the get commerce pages procedure is successful', async () => {
    getCommercePages.mockResolvedValueOnce(mockCommercePages);
    getRankedCommercePage.mockResolvedValueOnce(mockCommercePages);

    expect.assertions(6);

    await store.dispatch(action(commercePagesQuery));

    const actionResults = store.getActions();

    expect(warnDeprecatedMethod).toHaveBeenCalledTimes(1);
    expect(normalizeSpy).toHaveBeenCalledTimes(1);
    expect(getCommercePages).toHaveBeenCalledTimes(1);
    expect(getCommercePages).toHaveBeenCalledWith(
      commercePagesQuery,
      expectedConfig,
    );
    expect(store.getActions()).toEqual(
      expect.arrayContaining([
        expect.objectContaining(
          {
            type: actionTypes.GET_COMMERCE_PAGES_REQUEST,
          },
          {
            payload: expectedCommercePagesNormalizedPayload,
            type: actionTypes.GET_COMMERCE_PAGES_SUCCESS,
          },
        ),
      ]),
    );
    expect(
      find(actionResults, { type: actionTypes.GET_COMMERCE_PAGES_SUCCESS }),
    ).toMatchSnapshot('Get commerce pages payload');
  });

  it('commerce page with more than one page', async () => {
    getCommercePages.mockResolvedValue(mockCommercePagesWithPages);
    getRankedCommercePage.mockResolvedValueOnce(mockCommercePages);

    expect.assertions(7);

    await store.dispatch(action(commercePagesQuery));

    const actionResults = store.getActions();

    expect(warnDeprecatedMethod).toHaveBeenCalledTimes(1);
    expect(normalizeSpy).toHaveBeenCalledTimes(1);
    expect(getCommercePages).toHaveBeenCalledTimes(2);
    expect(getCommercePages).toHaveBeenCalledWith(
      commercePagesQuery,
      expectedConfig,
    );
    expect(getCommercePages).toHaveBeenLastCalledWith(
      { ...commercePagesQuery, PageIndex: 2 },
      expectedConfig,
    );
    expect(store.getActions()).toEqual(
      expect.arrayContaining([
        expect.objectContaining(
          {
            type: actionTypes.GET_COMMERCE_PAGES_REQUEST,
          },
          {
            payload: expectedCommercePagesNormalizedPayload,
            type: actionTypes.GET_COMMERCE_PAGES_SUCCESS,
          },
        ),
      ]),
    );
    expect(
      find(actionResults, { type: actionTypes.GET_COMMERCE_PAGES_SUCCESS }),
    ).toMatchSnapshot('Get commerce pages payload');
  });
});
