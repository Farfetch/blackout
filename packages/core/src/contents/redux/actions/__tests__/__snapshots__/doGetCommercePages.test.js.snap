// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doGetCommercePages() action creator commerce page with more than one page: Get commerce pages payload 1`] = `
Object {
  "meta": Object {
    "query": Object {
      "brand": 5030844,
      "category": 136643,
      "gender": 0,
      "type": "LISTING",
    },
  },
  "payload": Object {
    "entities": Object {
      "contentGroups": Object {
        "commerce_pages!woman": Object {
          "hash": "commerce_pages!woman",
        },
      },
    },
    "hash": "commerce_pages!woman",
    "result": "commerce_pages!woman",
  },
  "type": "@farfetch/blackout-core/GET_COMMERCE_PAGES_SUCCESS",
}
`;

exports[`doGetCommercePages() action creator should create the correct actions for when the get commerce pages procedure is successful: Get commerce pages payload 1`] = `
Object {
  "meta": Object {
    "query": Object {
      "brand": 5030844,
      "category": 136643,
      "gender": 0,
      "type": "LISTING",
    },
  },
  "payload": Object {
    "entities": Object {
      "contentGroups": Object {
        "commerce_pages!woman": Object {
          "hash": "commerce_pages!woman",
        },
      },
    },
    "hash": "commerce_pages!woman",
    "result": "commerce_pages!woman",
  },
  "type": "@farfetch/blackout-core/GET_COMMERCE_PAGES_SUCCESS",
}
`;
